docs/src/guide/testing.md

```diff
@@ -0,0 +1,345 @@
+
+Here we send an Insert event, which should start a timer. We filter out just the
+`Effect`s that were created by the `Timer` Capability, mapping them to their
+inner `Request<TimerOperation>` type.
+
+```rust,ignore,no_run
+let requests = &mut app
+    .update(Event::Insert("something".to_string()), &mut model)
+    .into_effects()
+    .filter_map(Effect::into_timer);
+```
+
+There are a few things to discuss here. Firstly, the `update()` method returns
+an `Update` struct, which contains vectors of `Event`s and `Effect`s. We are
+only interested in the `Effect`s, so we call `into_effects()` to consume them as
+an `Iterator` (there are also `effects()` and `effects_mut()` methods that allow
+us to borrow the `Effect`s instead of consuming them, but we don't need that
+here). Secondly, we use the `filter_map()` method to filter out just the
+`Effect`s that were created by the `Timer` Capability, using
+`Effect::into_timer` to map the `Effect`s to their inner
+`Request<TimerOperation>`.
+
+The [`Effect`](https://github.com/redbadger/crux/tree/master/crux_macros) derive
+macro generates filters and maps for each capability that we are using. So if
+our `Capabilities` struct looked like this...
+
+```rust,ignore,no_run
+
+#[cfg_attr(feature = "typegen", derive(crux_macros::Export))]
+#[derive(Effect)]
+#[effect(app = "NoteEditor")]
+pub struct Capabilities {
+    timer: Timer<Event>,
+    render: Render<Event>,
+    pub_sub: PubSub<Event>,
+    key_value: KeyValue<Event>,
+}
+```
+
+... we would get the following filters and filter_maps:
+
+```rust,ignore,no_run
+// filters
+Effect::is_timer(&self) -> bool
+Effect::is_render(&self) -> bool
+Effect::is_pub_sub(&self) -> bool
+Effect::is_key_value(&self) -> bool
+// filter_maps
+Effect::into_timer(self) -> Option<Request<TimerOperation>>
+Effect::into_render(self) -> Option<Request<RenderOperation>>
+Effect::into_pub_sub(self) -> Option<Request<PubSubOperation>>
+Effect::into_key_value(self) -> Option<Request<KeyValueOperation>>
+```
+
+We want to check that the first request is a `Start` operation, and that the
+timer is set to fire in 1000 milliseconds. The macro
+[`assert_let!()`](https://docs.rs/assert_let/0.1.0/assert_let/) does a pattern
+match for us and assigns the `id` to a local variable called `first_id`, which
+we'll use later. Finally, we don't expect any more timer requests to have been
+generated.
+
+```rust,ignore,no_run
+let mut request = requests.next().unwrap();
+assert_let!(
+    TimerOperation::Start {
+        id: first_id,
+        millis: 1000
+    },
+    request.operation.clone()
+);
+assert!(requests.next().is_none());
+```
+
+At this point the shell would start the timer — this is something the core can't
+do as it is a side effect — and so we need to tell the app that it was created.
+Note that resolving a request could call the app's `update()` resulting in more
+`Event`s being generated, which we need to feed back into the app.
```

Comment:
This makes me wonder whether we should have an API for this common pattern - something like `resolve_and_update`, but it's not immediately clear to me what it should do, there are a few scenarios... The `resolve` can potentially trigger multiple events _or_ another effect, so it might be two separate APIs...?

Either way, we might want to look for these patterns and provide some convenience shorthands.

Reply:
yep totally agree. For another PR though I think.